File: eval/extension/tools/list_files_tool.py
import os
import traceback
from eval.extension.tools.base_tool import BaseAgentTool

# Constants for file limits
LIST_FILES_LIMIT = 100  # Example value for the maximum number of files to return


class ListFilesTool(BaseAgentTool):
    def __init__(self, params, options):
        super().__init__(options)
        self.params = params

    async def execute(self):
        input_data = self.params.get("input", {})
        rel_dir_path = input_data.get("path")
        recursive_raw = input_data.get("recursive", "false")

        if rel_dir_path is None:
            return """
            Error: Missing value for required parameter 'path'. Please retry with complete response.
            A good example of a listFiles tool call is:
            {
                "tool": "list_files",
                "path": "path/to/directory"
            }
            Please try again with the correct path, you are not allowed to list files without a path.
            """

        try:
            recursive = recursive_raw.lower() == "true"
            absolute_path = os.path.abspath(os.path.join(self.cwd, rel_dir_path))
            files = await self.list_files(absolute_path, recursive)
            result = self.format_files_list(absolute_path, files)

            # If read-only mode is allowed, return the file list
            if self.always_allow_read_only:
                return result
            else:
                # No user prompt, returning the file list directly
                return result

        except Exception as error:
            error_string = (
                f"Error listing files and directories: {traceback.format_exc()}"
            )
            return error_string

    async def list_files(self, absolute_path, recursive):
        """Recursively list files in a directory."""
        file_list = []
        if recursive:
            for root, dirs, files in os.walk(absolute_path):
                for file_name in files:
                    file_list.append(os.path.join(root, file_name))
        else:
            try:
                file_list = [
                    os.path.join(absolute_path, f)
                    for f in os.listdir(absolute_path)
                    if os.path.isfile(os.path.join(absolute_path, f))
                ]
            except PermissionError:
                file_list = []
        return file_list

    def format_files_list(self, absolute_path, files):
        """Format the file list for output."""
        sorted_files = sorted(
            [os.path.relpath(file, absolute_path) for file in files],
            key=lambda x: (x.count(os.path.sep), x.lower()),
        )

        # Handling file truncation for large lists
        if len(sorted_files) >= LIST_FILES_LIMIT:
            truncated_list = "\n".join(sorted_files[:LIST_FILES_LIMIT])
            return f"{truncated_list}\n\n(Truncated at {LIST_FILES_LIMIT} results. Try listing files in subdirectories if you need to explore further.)"
        elif len(sorted_files) == 0:
            return (
                "No files found or you do not have permission to view this directory."
            )
        else:
            return "\n".join(sorted_files)


# Example usage (You can modify or remove this for your actual implementation)
options = {
    "cwd": "/path/to/directory",
    "alwaysAllowReadOnly": True,
    "alwaysAllowWriteOnly": False,
    "koduDev": None,
    "setRunningProcessId": None,
}
params = {"input": {"path": "example_directory", "recursive": "false"}}

tool = ListFilesTool(params, options)
result = tool.execute()

File: extension/src/agent/v1/tools/list-files.tool.ts
import * as path from "path"
import { serializeError } from "serialize-error"
import { LIST_FILES_LIMIT, listFiles } from "../../../parse-source-code"
import { ClaudeAsk, ClaudeSay, ClaudeSayTool } from "../../../shared/ExtensionMessage"
import { ToolResponse } from "../types"
import { formatGenericToolFeedback, formatToolResponse, getReadablePath } from "../utils"
import { AgentToolOptions, AgentToolParams } from "./types"
import { BaseAgentTool } from "./base-agent.tool"

export class ListFilesTool extends BaseAgentTool {
	protected params: AgentToolParams

	constructor(params: AgentToolParams, options: AgentToolOptions) {
		super(options)
		this.params = params
	}

	async execute(): Promise<ToolResponse> {
		const { input, ask, say } = this.params
		const { path: relDirPath, recursive: recursiveRaw } = input

		if (relDirPath === undefined) {
			await say(
				"error",
				"Claude tried to use list_files without value for required parameter 'path'. Retrying..."
			)
			return `Error: Missing value for required parameter 'path'. Please retry with complete response.
						A good example of a listFiles tool call is:
			{
				"tool": "list_files",
				"path": "path/to/directory"
			}
			Please try again with the correct path, you are not allowed to list files without a path.
			`
		}

		try {
			const recursive = recursiveRaw?.toLowerCase() === "true"
			const absolutePath = path.resolve(this.cwd, relDirPath)
			const files = await listFiles(absolutePath, recursive, 200)
			const result = this.formatFilesList(absolutePath, files[0])

			const message = JSON.stringify({
				tool: recursive ? "listFilesRecursive" : "listFilesTopLevel",
				path: getReadablePath(relDirPath, this.cwd),
				content: result,
			} as ClaudeSayTool)

			if (this.alwaysAllowReadOnly) {
				await say("tool", message)
			} else {
				const { response, text, images } = await ask("tool", message)

				if (response !== "yesButtonTapped") {
					if (response === "messageResponse") {
						await say("user_feedback", text, images)
						return formatToolResponse(formatGenericToolFeedback(text), images)
					}

					return "The user denied this operation."
				}
			}

			return result
		} catch (error) {
			const errorString = `Error listing files and directories: ${JSON.stringify(serializeError(error))}`
			await say(
				"error",
				`Error listing files and directories:\n${
					(error as Error).message ?? JSON.stringify(serializeError(error), null, 2)
				}`
			)

			return errorString
		}
	}

	async listFilesTopLevel(
		relDirPath: string | undefined,
		ask: (type: ClaudeAsk, question?: string) => Promise<{ response: string; text?: string; images?: string[] }>,
		say: (type: ClaudeSay, text?: string, images?: string[]) => void
	): Promise<ToolResponse> {
		if (relDirPath === undefined) {
			await say(
				"error",
				"Claude tried to use list_files_top_level without value for required parameter 'path'. Retrying..."
			)

			return `Error: Missing value for required parameter 'path'. Please retry with complete response.
			An example of a good listFilesTopLevel tool call is:
			{
				"tool": "list_files_top_level",
				"path": "path/to/directory"
			}
			Please try again with the correct path, you are not allowed to list files without a path.
			`
		}
		try {
			const absolutePath = path.resolve(this.cwd, relDirPath)
			const files = await listFiles(absolutePath, false, 200)
			const result = this.formatFilesList(absolutePath, files[0])

			const message = JSON.stringify({
				tool: "listFilesTopLevel",
				path: getReadablePath(relDirPath, this.cwd),
				content: result,
			} as ClaudeSayTool)
			if (this.alwaysAllowReadOnly) {
				await say("tool", message)
			} else {
				const { response, text, images } = await ask("tool", message)

				if (response !== "yesButtonTapped") {
					if (response === "messageResponse") {
						await say("user_feedback", text, images)
						return formatToolResponse(formatGenericToolFeedback(text), images)
					}

					return "The user denied this operation."
				}
			}

			return result
		} catch (error) {
			const errorString = `Error listing files and directories: ${JSON.stringify(serializeError(error))}`
			await say(
				"error",
				`Error listing files and directories:\n${
					(error as Error).message ?? JSON.stringify(serializeError(error), null, 2)
				}`
			)

			return errorString
		}
	}

	async listFilesRecursive(
		relDirPath: string | undefined,
		ask: (type: ClaudeAsk, question?: string) => Promise<{ response: string; text?: string; images?: string[] }>,
		say: (type: ClaudeSay, text?: string, images?: string[]) => void
	): Promise<ToolResponse> {
		if (relDirPath === undefined) {
			await say(
				"error",
				"Claude tried to use list_files_recursive without value for required parameter 'path'. Retrying..."
			)

			return `Error: Missing value for required parameter 'path'. Please retry with complete response.
			An example of a good listFilesRecursive tool call is:
			{
				"tool": "list_files_recursive",
				"path": "path/to/directory"
			}
			Please try again with the correct path, you are not allowed to list files without a path.
			`
		}

		try {
			const absolutePath = path.resolve(this.cwd, relDirPath)
			const files = await listFiles(absolutePath, true, 200)
			const result = this.formatFilesList(absolutePath, files[0])

			const message = JSON.stringify({
				tool: "listFilesRecursive",
				path: getReadablePath(relDirPath, this.cwd),
				content: result,
			} as ClaudeSayTool)
			if (this.alwaysAllowReadOnly) {
				await say("tool", message)
			} else {
				const { response, text, images } = await ask("tool", message)
				if (response !== "yesButtonTapped") {
					if (response === "messageResponse") {
						await say("user_feedback", text, images)
						return formatToolResponse(formatGenericToolFeedback(text), images)
					}

					return "The user denied this operation."
				}
			}

			return result
		} catch (error) {
			const errorString = `Error listing files recursively: ${JSON.stringify(serializeError(error))}`
			await say(
				"error",
				`Error listing files recursively:\n${
					(error as Error).message ?? JSON.stringify(serializeError(error), null, 2)
				}`
			)

			return errorString
		}
	}

	formatFilesList(absolutePath: string, files: string[]): string {
		const sorted = files
			.map((file) => {
				// convert absolute path to relative path
				const relativePath = path.relative(absolutePath, file)
				return file.endsWith("/") ? relativePath + "/" : relativePath
			})
			// Sort so files are listed under their respective directories to make it clear what files are children of what directories. Since we build file list top down, even if file list is truncated it will show directories that claude can then explore further.
			.sort((a, b) => {
				const aParts = a.split("/")
				const bParts = b.split("/")
				for (let i = 0; i < Math.min(aParts.length, bParts.length); i++) {
					if (aParts[i] !== bParts[i]) {
						// If one is a directory and the other isn't at this level, sort the directory first
						if (i + 1 === aParts.length && i + 1 < bParts.length) {
							return -1
						}
						if (i + 1 === bParts.length && i + 1 < aParts.length) {
							return 1
						}
						// Otherwise, sort alphabetically
						return aParts[i].localeCompare(bParts[i], undefined, { numeric: true, sensitivity: "base" })
					}
				}
				// If all parts are the same up to the length of the shorter path,
				// the shorter one comes first
				return aParts.length - bParts.length
			})
		if (sorted.length >= LIST_FILES_LIMIT) {
			const truncatedList = sorted.slice(0, LIST_FILES_LIMIT).join("\n")
			return `${truncatedList}\n\n(Truncated at ${LIST_FILES_LIMIT} results. Try listing files in subdirectories if you need to explore further.)`
		} else if (sorted.length === 0 || (sorted.length === 1 && sorted[0] === "")) {
			return "No files found or you do not have permission to view this directory."
		} else {
			return sorted.join("\n")
		}
	}
}
